content/blog/2017/2017-02-23-declarative-saucelabs-xunit.adoc

```diff
@@ -0,0 +1,399 @@
+    checkout scm
+
+    // Install dependencies
+    sh 'npm install'
+
+    stage "Test"
+    // Add sauce credentials
+    sauce('f0a6b8ad-ce30-4cba-bf9a-95afbc470a8a') {
+        // Start sauce connect
+        sauceconnect() {
+
+            // List of browser configs we'll be testing against.
+            def platform_configs = [
+                'chrome',
+                'firefox',
+                'ie',
+                'edge'
+            ].join(',')
+
+            // Nightwatch.js supports color ouput, so wrap this step for ansi color
+            wrap([$class: 'AnsiColorBuildWrapper', 'colorMapName': 'XTerm']) {
+                // Run selenium tests using Nightwatch.js
+                // Ignore error codes. The junit publisher will cover setting build status.
+                sh "./node_modules/.bin/nightwatch -e ${platform_configs} --test tests/guineaPig.js || true"
+            }
+
+            step([$class: 'XUnitBuilder',
+                thresholds: [
+                    [$class: 'SkippedThreshold', failureThreshold: '0'],
+                    // Allow for a significant number of failures
+                    // Keeping this threshold so that overwhelming failures are guaranteed
+                    //     to still fail the build
+                    [$class: 'FailedThreshold', failureThreshold: '10']],
+                tools: [[$class: 'JUnitType', pattern: 'reports/**']]])
+
+            saucePublisher()
+        }
+    }
+}
+----
+
+image::/images/post-images/2017-02-23/blue-ocean.png[Blue Ocean Run, role="center", width=800]
+
+image::/images/post-images/2017-02-23/sauce-report.png[SauceLabs Test Report, role="center", width=800]
+
+NOTE: Blue Ocean doesn't support displaying SauceLabs test reports yet
+(see link:https://issues.jenkins-ci.org/browse/JENKINS-42242[JENKINS-42242]).
+To get view the report above, I navigated back down to the stage view of this run.
+
+== Elevating Settings using `environment`
+
+Each time we've moved a project from Scripted Pipeline to Declarative,
+we've found the cleaner format of Declarative Pipeline highlights the less
+clear parts of the existing Pipeline.
+In this case, the first thing that jumps out at me is that the parameters of the
+Saucelabs and Nightwatch execution are hardcoded and buried down in the middle of our Pipeline.
+This is a relatively short Pipeline, so it isn't terribly hard to find them,
+but as this pipeline grows and changes it would be better if those values were kept separate.
+In Scripted, we'd have defined some variables,
+but Declarative doesn't allow us to define variables in the usual Groovy sense.
+
+The `environment` directive let's us set some environment variables
+and use them later in our pipeline.
+As you'd expect, the `environment` directive is just a set of name-value pairs.
+Environment variables are accessible in Pipeline via `env.variableName` (or just `variableName`)
+and in shell scripts as standard environment variables, `$variableName`.
```

Comment:
Perhaps phrase as *typically `$variableName`* for crazy people using weird shells with different syntax?